// woocommerce-rest-api/wc.js
'use strict';
/**
 * A WC REST API client for Node.js
 *
 * @example
 *     var wc = new WC({ endpoint: 'http://src.wordpress-develop.dev/wc-api/v1/' });
 *     wc.orders().then(function( orders ) {
 *         console.log( orders );
 *     }).catch(function( err ) {
 *         console.error( err );
 *     });
 *
 * @module WC
 * @main WC
 * @beta
 })
 */
var extend = require( 'node.extend' );

var defaults = {
	username: '',
	password: ''
};

// Pull in request module constructors
var OrdersRequest = require( './lib/orders' );
var CollectionRequest = require( './lib/shared/collection-request' );
var WCRequest = require( './lib/shared/wc-request' );

/**
 * The base constructor for the WC API service
 *
 * @class WC
 * @constructor
 * @uses PostsRequest
 * @uses TaxonomiesRequest
 * @uses UsersRequest
 * @param {Object} options An options hash to configure the instance
 * @param {String} options.endpoint The URI for a WC-API endpoint
 * @param {String} [options.username] A WC-API Basic Auth username
 * @param {String} [options.password] A WC-API Basic Auth password
 */
function WC( options ) {

	// Enforce `new`
	if ( this instanceof WC === false ) {
		return new WC( options );
	}

	this._options = extend( {}, defaults, options );

	if ( ! this._options.endpoint ) {
		throw new Error( 'options hash must contain an API endpoint URL string' );
	}

	// Ensure trailing slash on endpoint URI
	this._options.endpoint = this._options.endpoint.replace( /\/?$/, '/' );

	return this;
}

/**
 * Convenience method for making a new WC instance
 *
 * @example
 * These are equivalent:
 *
 *     var wc = new WC({ endpoint: 'http://my.blog.url/wc-api/v1/' });
 *     var wc = WC.site( 'http://my.blog.url/wc-api/v1/' );
 *
 * @method site
 * @static
 * @param {String} endpoint The URI for a WC-API endpoint
 * @return {WC} A new WC instance, bound to the provided endpoint
 */
WC.site = function( endpoint ) {
	return new WC({ endpoint: endpoint });
};

/**
 * Start a request against the `/media` endpoint
 *
 * @method media
 * @param {Object} [options] An options hash for a new MediaRequest
 * @return {MediaRequest} A MediaRequest instance
 */
WC.prototype.media = function( options ) {
	options = options || {};
	options = extend( options, this._options );
	return new MediaRequest( options );
};

/**
 * Start a request against the `/pages` endpoint
 *
 * @method pages
 * @param {Object} [options] An options hash for a new PagesRequest
 * @return {PagesRequest} A PagesRequest instance
 */
WC.prototype.pages = function( options ) {
	options = options || {};
	options = extend( options, this._options );
	return new PagesRequest( options );
};

/**
 * Start a request against the `/orders` endpoint
 *
 * @method orders
 * @param {Object} [options] An options hash for a new OrdersRequest
 * @return {OrdersRequest} A OrdersRequest instance
 */
WC.prototype.orders = function( options ) {
	options = options || {};
	options = extend( options, this._options );
	return new OrdersRequest( options );
};

/**
 * Start a request against the `taxonomies` endpoint
 *
 * @method taxonomies
 * @param {Object} [options] An options hash for a new TaxonomiesRequest
 * @return {TaxonomiesRequest} A TaxonomiesRequest instance
 */
WC.prototype.taxonomies = function( options ) {
	options = options || {};
	options = extend( options, this._options );
	return new TaxonomiesRequest( options );
};

/**
 * Start a request for a specific taxonomy object
 *
 * It is repetitive to have to type `.taxonomies().taxonomy()` whenever you want to request
 * a taxonomy object or list of terms for a taxonomy. This convenience method lets you
 * create a `TaxonomiesRequest` object that is bound to the provided taxonomy name.
 *
 * @example
 * If your site uses two custom taxonomies, book_genre and book_publisher, before you would
 * have had to request these terms using the verbose form:
 *
 *     wc.taxonomies().taxonomy( 'book_genre' ).terms()...
 *     wc.taxonomies().taxonomy( 'book_publisher' )...
 *
 * Using `.taxonomy()`, the same query can be achieved much more succinctly:
 *
 *     wc.taxonomy( 'book_genre' ).terms()...
 *     wc.taxonomy( 'book_publisher' )...
 *
 * @method taxonomy
 * @param {String} taxonomyName The name of the taxonomy to request
 * @return {TaxonomiesRequest} A TaxonomiesRequest object bound to the value of taxonomyName
 */
WC.prototype.taxonomy = function( taxonomyName ) {
	var options = extend( {}, this._options );
	return new TaxonomiesRequest( options ).taxonomy( taxonomyName );
};

/**
 * Request a list of category terms
 *
 * This is a shortcut method to retrieve the terms for the "category" taxonomy
 *
 * @example
 * These are equivalent:
 *
 *     wc.taxonomies().taxonomy( 'category' ).terms()
 *     wc.categories()
 *
 * @method categories
 * @return {TaxonomiesRequest} A TaxonomiesRequest object bound to the terms for "category"
 */
WC.prototype.categories = function() {
	var options = extend( {}, this._options );
	return new TaxonomiesRequest( options ).taxonomy( 'category' ).terms();
};

/**
 * Request a list of post_tag terms
 *
 * This is a shortcut method to retrieve the terms for the "post_tag" taxonomy
 *
 * @example
 * These are equivalent:
 *
 *     wc.taxonomies().taxonomy( 'post_tag' ).terms()
 *     wc.tags()
 *
 * @method tags
 * @return {TaxonomiesRequest} A TaxonomiesRequest object bound to the terms for "post_tag"
 */
WC.prototype.tags = function() {
	var options = extend( {}, this._options );
	return new TaxonomiesRequest( options ).taxonomy( 'post_tag' ).terms();
};

/**
 * Start a request against the `/posts/types` endpoint
 *
 * @method types
 * @param {Object} [options] An options hash for a new TypesRequest
 * @return {TypesRequest} A TypesRequest instance
 */
WC.prototype.types = function( options ) {
	options = options || {};
	options = extend( options, this._options );
	return new TypesRequest( options );
};

/**
 * Start a request against the `/users` endpoint
 *
 * @method users
 * @param {Object} [options] An options hash for a new UsersRequest
 * @return {UsersRequest} A UsersRequest instance
 */
WC.prototype.users = function( options ) {
	options = options || {};
	options = extend( options, this._options );
	return new UsersRequest( options );
};

/**
 * Define a method to handle specific custom post types.
 *
 * @example
 * If your site had an events custom type with name `event_cpt`, you could create a convenience
 * method for querying events and store it on the WC instance.
 *
 * Create the WC instance, define the custom endpoint handler, and save it to `wc.events`:
 *
 *     var wc = new WC({ endpoint: 'http://some-website.com/wc-api/v1/' });
 *     wc.events = wc.registerType( 'event_cpt' );
 *
 * You can now call `wc.events()` to trigger event post requests
 *
 *     wc.events().get()... // equivalent to wc.posts().type( 'event_cpt' ).get()...
 *
 * `registerType()` just returns a function, so there's no requirement to store it as a property
 * on the WC instance; however, following the above pattern is likely to be the most useful.
 *
 * @method registerType
 * @param {String|Array} type A string or array of post type names
 * @return {Function} A function to create PostsRequests pre-bound to the provided types
 */
WC.prototype.registerType = function( type ) {
	var options = extend( {}, this._options );
	return function() {
		return new PostsRequest( options ).type( type );
	};
};

/**
 * Generate a request against a completely arbitrary endpoint, with no assumptions about
 * or mutation of path, filtering, or query parameters. This request is not restricted to
 * the endpoint specified during WC object instantiation.
 *
 * @example
 * Generate a request to the explicit URL "http://your.website.com/wc-api/v1/some/custom/path" (yeah, we wish ;)
 *
 *     wc.url( 'http://your.website.com/wc-api/v1/some/custom/path' ).get()...
 *
 * @method url
 * @param {String} url The URL to request
 * @return {WCRequest} A WCRequest object bound to the provided URL
 */
WC.prototype.url = function( url ) {
	var options = extend( {}, this._options, {
		endpoint: url
	});
	return new WCRequest( options );
};

/**
 * Generate a query against an arbitrary path on the current endpoint. This is useful for
 * requesting resources at custom WC-API endpoints, such as WooCommerce's `/products`.
 *
 * @method root
 * @param {String} [relativePath] An endpoint-relative path to which to bind the request
 * @param {Boolean} [collection] Whether to return a CollectionRequest or a vanilla WCRequest
 * @return {CollectionRequest|WCRequest} A request object
 */
WC.prototype.root = function( relativePath, collection ) {
	relativePath = relativePath || '';
	collection = collection || false;
	var options = extend( {}, this._options );
	// Request should be
	var request = collection ? new CollectionRequest( options ) : new WCRequest( options );

	// Set the path template to the string passed in
	request._template = relativePath;

	return request;
};

module.exports = WC;
