// woocommerce-rest-api/lib/orders.js
'use strict';
/**
 * @module WC
 * @submodule OrdersRequest
 * @beta
 */
var CollectionRequest = require( './shared/collection-request' );
var inherit = require( 'util' ).inherits;

/**
 * OrdersRequest extends CollectionRequest to handle the /orders API endpoint
 *
 * @class OrdersRequest
 * @constructor
 * @extends CollectionRequest
 * @param {Object} options A hash of options for the OrdersRequest instance
 * @param {String} options.endpoint The endpoint URI for the invoking WC instance
 * @param {String} [options.username] A username for authenticating API requests
 * @param {String} [options.password] A password for authenticating API requests
 */
function OrdersRequest( options ) {
	/**
	 * Configuration options for the request such as the endpoint for the invoking WC instance
	 * @property _options
	 * @type Object
	 * @private
	 * @default {}
	 */
	this._options = options || {};

	/**
	 * A hash of filter values to parse into the final request URI
	 * @property _filters
	 * @type Object
	 * @private
	 * @default {}
	 */
	this._filters = {};

	/**
	 * A hash of taxonomy terms to parse into the final request URI
	 * @property _taxonomyFilters
	 * @type Object
	 * @private
	 * @default {}
	 */
	this._taxonomyFilters = {};

	/**
	 * A hash of non-filter query parameters
	 *
	 * @property _params
	 * @type Object
	 * @private
	 * @default {}
	 */
	this._params = {};

	/**
	 * A hash of values to assemble into the API request path
	 *
	 * @property _path
	 * @type Object
	 * @private
	 * @default {}
	 */
	this._path = {};

	/**
	 * The URL template that will be used to assemble endpoint paths
	 *
	 * @property _template
	 * @type String
	 * @private
	 * @default 'orders(/:id)(/:action)(/:actionId)'
	 */
	this._template = 'orders(/:id)(/:action)(/:actionId)';

	/**
	 * @property _supportedMethods
	 * @type Array
	 * @private
	 * @default [ 'head', 'get', 'post' ]
	 */
	this._supportedMethods = [ 'head', 'get', 'post' ];
}

inherit( OrdersRequest, CollectionRequest );

/**
 * A hash table of path keys and regex validators for those path elements
 *
 * @property _pathValidators
 * @type Object
 * @private
 */
OrdersRequest.prototype._pathValidators = {

	/**
	 * ID must be an integer
	 *
	 * @property _pathValidators.id
	 * @type {RegExp}
	 */
	id: /^\d+$/,

	/**
	 * Action must be one of 'meta', 'comments', or 'revisions'
	 *
	 * @property _pathValidators.action
	 * @type {RegExp}
	 */
	action: /(meta|comments|revisions)/
};

/**
 * Specify a post ID to query
 *
 * @method id
 * @chainable
 * @param {Number} id The ID of a post to retrieve
 * @return {OrdersRequest} The OrdersRequest instance (for chaining)
 */
OrdersRequest.prototype.id = function( id ) {
	this._path.id = parseInt( id, 10 );
	this._supportedMethods = [ 'head', 'get', 'put', 'post', 'delete' ];

	return this;
};

/**
 * Specify that we are getting the comments for a specific post
 *
 * @method comments
 * @chainable
 * @return {OrdersRequest} The OrdersRequest instance (for chaining)
 */
OrdersRequest.prototype.comments = function() {
	this._path.action = 'comments';
	this._supportedMethods = [ 'head', 'get', 'post' ];
	return this;
};

/**
 * Specify a particular comment to retrieve
 * (forces action "comments")
 *
 * @method comment
 * @chainable
 * @param {Number} id The ID of the comment to retrieve
 * @return {OrdersRequest}
 */
OrdersRequest.prototype.comment = function( id ) {
	this._path.action = 'comments';
	this._path.actionId = parseInt( id, 10 );
	this._supportedMethods = [ 'head', 'get', 'delete' ];

	return this;
};

/**
 * Query a collection of orders for orders of a specific type
 *
 * @method type
 * @param {String|Array} type A string or array of strings specifying post types to query
 * @chainable
 * @return {OrdersRequest} The OrdersRequest instance (for chaining)
 */
OrdersRequest.prototype.type = function( type ) {
	this.param( 'type', type, true );
	return this;
};

/**
 * Specify that we are requesting the revisions for a specific post (forces basic auth)
 *
 * @method revisions
 * @chainable
 * @return {OrdersRequest} The OrdersRequest instance (for chaining)
 */
OrdersRequest.prototype.revisions = function() {
	this._path.action = 'revisions';
	this._supportedMethods = [ 'head', 'get' ];

	return this.auth();
};

/**
 * @method statuses
 * @chainable
 * @return {OrdersRequest} The OrdersRequest instance (for chaining)
 */
OrdersRequest.prototype.statuses = function() {
	this._path.action = 'statuses';
	this._supportedMethods = [ 'head', 'get' ];

	return this;
};

module.exports = OrdersRequest;
